Retry when DynamoDB source fails to initialize due to permission issue

**Is your feature request related to a problem? Please describe.**
Currently, if DDB source role does not have `dynamodb:DescribeTable` permission, the pipeline will shut down during initialization. We can catch this as a retryable error like we did for Opensearch sink to allow user to correct permissions on the role.

The permission error is like this:
```
[dynamodb-pipeline-sink-worker-2-thread-1] ERROR org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor - Pipeline [dynamodb-pipeline] process worker encountered a fatal exception, cannot proceed further
java.util.concurrent.ExecutionException: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::123456789012:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:123456789012:table/test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
	at org.opensearch.dataprepper.pipeline.common.PipelineThreadPoolExecutor.afterExecute(PipelineThreadPoolExecutor.java:70) ~[data-prepper-core-2.6.0-SNAPSHOT.jar:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1129) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: software.amazon.awssdk.services.dynamodb.model.DynamoDbException: User: arn:aws:sts::193777858833:assumed-role/pipelineRole/Data-Prepper-0233fade-6af1-4da2-a36b-923d9802c8cb is not authorized to perform: dynamodb:DescribeTable on resource: arn:aws:dynamodb:us-east-1:193777858833:table/gameday-test-source-table because no identity-based policy allows the dynamodb:DescribeTable action (Service: DynamoDb, Status Code: 400, Request ID: 839GJP2AL1PSOC8225BKVCULM3VV4KQNSO5AEMVJF66Q9ASUAAJG)
	at software.amazon.awssdk.core.internal.http.CombinedResponseHandler.handleErrorResponse(CombinedResponseHandler.java:125) ~[sdk-core-2.20.67.jar:?]
	...
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56) ~[aws-core-2.20.67.jar:?]
	at software.amazon.awssdk.services.dynamodb.DefaultDynamoDbClient.describeTable(DefaultDynamoDbClient.java:2344) ~[dynamodb-2.20.67.jar:?]
	at org.opensearch.dataprepper.plugins.source.dynamodb.DynamoDBService.getTableInfo(DynamoDBService.java:271) ~[dynamodb-source-2.6.0-SNAPSHOT.jar:?]
...
```

**Describe the solution you'd like**
A clear and concise description of what you want to happen.

**Describe alternatives you've considered (Optional)**
A clear and concise description of any alternative solutions or features you've considered.

**Additional context**
Add any other context or screenshots about the feature request here.

--- a/data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java
+++ b/data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/leader/LeaderScheduler.java
@@ -77,40 +77,42 @@ public class LeaderScheduler implements Runnable {
                     if (sourcePartition.isPresent()) {
                         LOG.info("Running as a LEADER node");
                         leaderPartition = (LeaderPartition) sourcePartition.get();
-
-                        LeaderProgressState leaderProgressState = leaderPartition.getProgressState().get();
-                        if (!leaderProgressState.isInitialized()) {
-                            LOG.debug("The service is not been initialized");
-                            init();
-                        }
-
+                    }
+                }
+                // Once owned, run Normal LEADER node process.
+                // May want to quit this scheduler if streaming is not required
+                if (leaderPartition != null) {
+                    LeaderProgressState leaderProgressState = leaderPartition.getProgressState().get();
+                    if (!leaderProgressState.isInitialized()) {
+                        LOG.debug("The service is not been initialized");
+                        init();
+                    } else {
                         // The initialization process will populate that value, otherwise, get from state
                         if (streamArns == null) {
                             streamArns = leaderProgressState.getStreamArns();
                         }
                     }
-                }
-                // Once owned, run Normal LEADER node process.
-                // May want to quit this scheduler if streaming is not required
-                if (leaderPartition != null && !streamArns.isEmpty()) {
 
-                    // Step 1: Run shard discovery
-                    streamArns.forEach(streamArn -> {
-                        shardManager.runDiscovery(streamArn);
-                    });
+                    if (streamArns != null && !streamArns.isEmpty()) {
+                        // Step 1: Run shard discovery
+                        streamArns.forEach(streamArn -> {
+                            shardManager.runDiscovery(streamArn);
+                        });
 
-                    // Step 2: Search all completed shards in the last 1 day (maximum time)
-                    List<EnhancedSourcePartition> sourcePartitions = coordinator.queryCompletedPartitions(
-                            StreamPartition.PARTITION_TYPE,
-                            Instant.now().minus(Duration.ofDays(1))
-                    );
+                        // Step 2: Search all completed shards in the last 1 day (maximum time)
+                        List<EnhancedSourcePartition> sourcePartitions = coordinator.queryCompletedPartitions(
+                                StreamPartition.PARTITION_TYPE,
+                                Instant.now().minus(Duration.ofDays(1))
+                        );
 
-                    // Step 3: Find and create children partitions.
-                    compareAndCreateChildrenPartitions(sourcePartitions);
+                        // Step 3: Find and create children partitions.
+                        compareAndCreateChildrenPartitions(sourcePartitions);
 
-                    // Extend the timeout
-                    // will always be a leader until shutdown
-                    coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+                        // Extend the timeout
+                        // will always be a leader until shutdown
+                        coordinator.saveProgressStateForPartition(leaderPartition, Duration.ofMinutes(DEFAULT_EXTEND_LEASE_MINUTES));
+                    }
+
                 }
 
             } catch (Exception e) {

